### PR TITLE
Support SymInt steps for linspace/logspace export

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184939

`torch.export` can represent data-dependent tensor scalars as unbacked
SymInts when the user provides the needed runtime checks. That already works
for shape-producing operators such as `arange`, but `linspace` and `logspace`
still declared `steps` as a plain `int`. Passing a tensor-derived value for
`steps` therefore forced argument binding to specialize the symbol before fake
or meta handling could preserve it, raising GuardOnDataDependentSymNode.

Change the public operator schemas for `linspace` and `logspace` so `steps` is
a `SymInt`, matching its role as the output length. Update the functorch batch
rule signatures for the same schema change; those eager batch rules still guard
to a concrete value internally because they materialize the sequence directly.
Add an export regression test that constrains a tensor-derived step count and
checks both strict and non-strict export.

Fixes #172930
Generated by my agent

Test Plan:
- PYTHONPATH=/data/users/jansel/pytorch-issue-fixer/pytorch ninja -C build torch_python
- python test/export/test_export.py TestExport.test_linspace_logspace_unbacked_steps
- python test/test_tensor_creation_ops.py -k linspace -q
- python test/test_tensor_creation_ops.py -k logspace -q
- lintrunner -a
- git diff --cached --check

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says Dr. CI reports only unrelated unstable failures, so I added the release notes: export label, marked this ready for review, and requested reviewers.

**Review by oulgen — APPROVED** [review]

File: aten/src/ATen/functorch/BatchRulesFactory.cpp [hunk 1]
```diff
@@ -121,14 +121,15 @@ static std::tuple<Tensor, std::optional<int64_t>> linspace_logspace_batch_rule_h
   end_ = moveBatchDimToFront(end_, end_bdim);
 
   auto tensor_options = at::TensorOptions().dtype(dtype).layout(layout).device(device).pinned_memory(pin_memory);
+  auto steps_int = steps.guard_int(__FILE__, __LINE__);
```
Comment (jansel) [inline aten/src/ATen/functorch/BatchRulesFactory.cpp]: Will this cause recompiles?  Agent, make sure the tests ensure this value can change without recompiling.

Comment (jansel) [reply aten/src/ATen/functorch/BatchRulesFactory.cpp]: My agent says I removed the concrete guard from the linspace/logspace vmap batch rule, kept the implementation on symbolic SymInt operations, and added compile-counter coverage showing exported modules and vmap both reuse one compiled frame when steps changes.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says CI was failing in test_linspace_logspace_unbacked_steps_strict_export_v2. I fixed the vmap batch rule dtype for steps=0 so it matches eager linspace/logspace, and updated the negative runtime assertion to accept strict-export-v2 guard failures.

**Comment by jansel** [comment]
@claude are those CI failures related to this PR?

**Comment by jansel** [comment]
My agent says I updated this PR to address the failing macOS vmap OpInfo CI for linspace. The failure was caused by the SymInt batch-rule rewrite letting the OpInfo error case become scalar after vmap slicing; the update keeps non-scalar error coverage and makes the linspace/logspace vmap helper mirror eager scalar validation, dtype inference, and complex dtype checks. I also preserved the concrete steps == 0 and steps == 1 paths while keeping the symbolic SymInt path for dynamic steps, which avoids precision loss and unsupported complex arange cases.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
More details on revert https://metamate.internalmeta.com/artifact/late-mustang-13

This PR should land after MTIA update: D108900774

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
